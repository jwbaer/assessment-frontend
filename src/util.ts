import { Component, List, PageApiResult, PageSpec, Variable, VariableDeclaration } from "./types";

export const convertPageApiResultToPageSpec = (apiResult: PageApiResult): PageSpec => {
    const componentIndex: Map<number, Component> = new Map(
        apiResult.components.map((obj: Component) => [obj.id, obj])
    );
    const listIndex: Map<number, List> = new Map(
        apiResult.lists.map((obj: List) => [obj.id, obj])
    );
    const variableIndex: Map<string, Variable> | undefined = apiResult.variables
        ? new Map(
            apiResult.variables.map((obj: VariableDeclaration) => [
                obj.name,
                { name: obj.name, type: obj.type, value: obj.initialValue },
            ])
        )
        : undefined;

    return({ componentIndex, listIndex, variableIndex });
}
