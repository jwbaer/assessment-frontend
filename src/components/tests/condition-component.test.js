import { queryByAltText, render, screen } from "@testing-library/react";

import ComponentPanel from "../component-panel";
import ConditionComponent from "../condition-component";
import PageContext from "../../contexts/page-context";
import { convertPageApiResultToPageSpec } from "../../util";

describe("<ConditionComponent /> spec", () => {
    const apiResult = {
        variables: [
            {
                name: "show_image",
                type: "string",
                initialValue: "hide",
            },
        ],
        lists: [
            {
                id: 0,
                components: [1],
            },
            {
                id: 1,
                components: [2],
            },
        ],
        components: [
            {
                id: 1,
                type: "condition",
                type: "condition",
                options: {
                    variable: "show_image",
                    value: "show",
                },
                children: 1,
            },
            {
                id: 2,
                type: "image",
                options: {
                    src: "/locations/new-york.png",
                    alt: "Cartoon of New York skyline",
                },
            },
        ],
    };

    const pageSpec = convertPageApiResultToPageSpec(apiResult);

    it("does not render the child list when condition is false", () => {
        const setPageSpec = jest.fn();
        const { asFragment } = render(
            <PageContext.Provider value={{ pageSpec, setPageSpec }}>
                <ConditionComponent component={pageSpec.componentIndex.get(1)} />
            </PageContext.Provider>
        );

        expect(queryByAltText(asFragment(), "Cartoon of New York skyline")).toBeFalsy();
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders the child list when condition is true", () => {
        pageSpec.variableIndex.get("show_image").value = "show";
        const setPageSpec = jest.fn();
        const { asFragment } = render(
            <PageContext.Provider value={{ pageSpec, setPageSpec }}>
                <ConditionComponent component={pageSpec.componentIndex.get(1)} />
            </PageContext.Provider>
        );

        const image = screen.getByAltText("Cartoon of New York skyline");
        expect(image).toBeDefined();
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders when specifed through the component panel", () => {
        pageSpec.variableIndex.get("show_image").value = "show";
        const setPageSpec = jest.fn();
        const { asFragment } = render(
            <PageContext.Provider value={{ pageSpec, setPageSpec }}>
                <ComponentPanel componentId={1} />
            </PageContext.Provider>
        );

        const image = screen.getByAltText("Cartoon of New York skyline");
        expect(image).toBeDefined();
        expect(asFragment()).toMatchSnapshot();
    });
});
