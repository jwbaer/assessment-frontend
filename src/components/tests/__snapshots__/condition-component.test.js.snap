// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConditionComponent /> spec does not render the child list when condition is false 1`] = `<DocumentFragment />`;

exports[`<ConditionComponent /> spec renders the child list when condition is true 1`] = `
<DocumentFragment>
  <div
    class="list"
  >
    <div
      class="component image"
      data-testid="image-component-2"
    >
      <img
        alt="Cartoon of New York skyline"
        data-testid="image-component-2-image"
        src="/locations/new-york.png"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ConditionComponent /> spec renders when specifed through the component panel 1`] = `
<DocumentFragment>
  <div
    class="list"
  >
    <div
      class="component image"
      data-testid="image-component-2"
    >
      <img
        alt="Cartoon of New York skyline"
        data-testid="image-component-2-image"
        src="/locations/new-york.png"
      />
    </div>
  </div>
</DocumentFragment>
`;
